### PR TITLE
Create AbsoluteFilePathLocator service

### LINKED ISSUE #1874: Create file path translation service
In order to support #1763, the relative path that we get back from the server should be converted to an absolute file path on disk.

### PR DESCRIPTION
**Pull request by rita-g-sonarsource**
Fixes #1874

### CODE REVIEW

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 1]
```diff
@@ -56,24 +56,19 @@ internal class ProjectSystemHelper : IProjectSystemHelper
...
         public ProjectSystemHelper(IServiceProvider serviceProvider)
         {
-            if (serviceProvider == null)
-            {
-                throw new ArgumentNullException(nameof(serviceProvider));
-            }
+            this.serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
 
-            this.serviceProvider = serviceProvider;
+            solution = this.serviceProvider.GetService<SVsSolution, IVsSolution>();
```
Comment (rita-g-sonarsource) [inline src/Integration/LocalServices/ProjectSystemHelper.cs:65]: I decided to fetch this in the ctor, so `EnumerateProjects` wouldn't have to get parameters

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 2]
```diff
@@ -455,7 +450,7 @@ private string GetSolutionItemsFolderName()
             return solutionItemsFolderName;
         }
 
-        private static IEnumerable<IVsHierarchy> EnumerateProjects(IVsSolution solution)
+        public IEnumerable<IVsHierarchy> EnumerateProjects()
```
Comment (rita-g-sonarsource) [inline src/Integration/LocalServices/ProjectSystemHelper.cs:453]: I'm using this now in the new service

File: src/TestInfrastructure/Framework/ConfigurableVsProjectSystemHelper.cs [hunk 3]
```diff
@@ -197,6 +198,21 @@ public bool IsLegacyProjectSystem(Project dteProject)
             return this.isLegacyProjectSystem;
         }
 
+        public IEnumerable<IVsHierarchy> EnumerateProjects()
```
Comment (rita-g-sonarsource) [inline src/TestInfrastructure/Framework/ConfigurableVsProjectSystemHelper.cs:201]: These aren't used in a test context

File: src/Integration.UnitTests/LocalServices/ProjectSystemHelperTests.cs [hunk 4]
```diff
@@ -52,7 +52,7 @@ public void TestInitialize()
         #region Tests
 
         [TestMethod]
-        public void ProjectSystemHelper_ArgCheck()
```
Comment (rita-g-sonarsource) [inline src/Integration.UnitTests/LocalServices/ProjectSystemHelperTests.cs:55]: Misc: I removed the `ProjectSystemHelper_` prefix from the tests

File: src/Core/IAbsoluteFilePathLocator.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarLint.VisualStudio.Core
+{
+    public interface IAbsoluteFilePathLocator
+    {
+        /// <summary>
+        /// Searches for all files in the current solution and returns an absolute file path that ends with the given relative path
```
Comment (duncanp-sonar) [inline src/Core/IAbsoluteFilePathLocator.cs:26]: I think it's worth calling out that the first match will be returned e.g.

> Returns the absolute path of the first file in the solution that ends with the given relative path

File: src/Integration/Helpers/AbsoluteFilePathLocator.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        public string Locate(string relativeFilePath)
+        {
+            if (relativeFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(relativeFilePath));
+            }
+
+            foreach (var vsHierarchy in projectSystemHelper.EnumerateProjects())
```
Comment (duncanp-sonar) [inline src/Integration/Helpers/AbsoluteFilePathLocator.cs:53]: Does this need to be done on the UI thread?

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 7]
```diff
@@ -502,5 +492,47 @@ public bool IsLegacyProjectSystem(Project dteProject)
...
+
+        private static IEnumerable<VSConstants.VSITEMID> ChildrenOf(IVsHierarchy hierarchy, uint rootId)
+        {
+            var result = new List<VSConstants.VSITEMID>();
+
+            for (var itemId = FirstChild(hierarchy, rootId);
+                itemId != (uint)VSConstants.VSITEMID.Nil;
+                itemId = NextSibling(hierarchy, itemId))
+            {
+                result.Add((VSConstants.VSITEMID) itemId);
```
Comment (duncanp-sonar) [inline src/Integration/LocalServices/ProjectSystemHelper.cs:510]: Did you decide against making this enumeration lazy?

Comment (duncanp-sonar) [reply src/Integration/LocalServices/ProjectSystemHelper.cs:510]: Ah - so the top-level projects are enumerated lazily, but all of the items in a project are fetched at once?
It's worth documenting this as it's not obvious.

**Review by duncanp-sonar — APPROVED** [review]

File: src/Integration/Helpers/AbsoluteFilePathLocator.cs (hunk 6 above)
Comment (rita-g-sonarsource) [reply src/Integration/Helpers/AbsoluteFilePathLocator.cs:53]: Doesn't seem to be so; there are no warnings, and I ran it on a background thread and was able to access the objects without errors.

File: src/Integration/LocalServices/ProjectSystemHelper.cs (hunk 7 above)
Comment (rita-g-sonarsource) [reply src/Integration/LocalServices/ProjectSystemHelper.cs:510]: I added a comment 👍

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 8]
```diff
@@ -502,5 +492,47 @@ public bool IsLegacyProjectSystem(Project dteProject)
...
+        private static uint GetItemId(IVsHierarchy hierarchy, uint itemId, __VSHPROPID propId)
+        {
+            hierarchy.GetProperty(itemId, (int)propId, out var idObj);
+
+            return idObj == null ? (uint)VSConstants.VSITEMID.Nil : (uint)(int) idObj;
+        }
+
+        public string GetItemFilePath(IVsProject vsProject, VSConstants.VSITEMID itemId)
+        {
+            var hr = vsProject.GetMkDocument((uint)itemId, out var filePath);
```
Comment (rita-g-sonarsource) [inline src/Integration/LocalServices/ProjectSystemHelper.cs:532]: @duncanp-sonar I did some testing in a c++ project and I get a debug exception here.
It seems to be that in c++ projects, `GetAllItems` returns also all included files, such as framework `C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\shared\basetsd.h`. I get hundreds of files in my tiny project, whilst with DTE I get only the relevant ones. I'm not sure how to fix `GetAllItems` to ignore all the c++ dependencies. I'll look into it and won't merge this PR for now.

**Pull request closed without merging**